doc/design/backfill-starvation-prevention.md

```diff
@@ -0,0 +1,179 @@
+# Backfill and Starvation Prevention
+
+## Motivation
+
+### Backfill
+The following diagrams illustrates why we need to add backfill capability to kube batch scheduler. 
+
+We assume that task queue contains 5 tasks, A1, A2, B, C, and D, among which A1 and A2 belong to job A. We further assume that A1 and A2 each requires 1 CPU, and A1 and A2 have to start at the same time. 
+
+Suppose the cluster has only one node. The node has two CPUs. CPU1 is idle and CPU2 is occupied. In the first round, scheduler allocates CPU1 to task A1. There is no CPU available for task A2. Since A1 and A2 have to start at the same time, A1 has to wait, even though it is allocated with enough CPU. Consequently, CPU1 will remain idle in the first round which is a waste of resource. One objective of this feature is to improve resource utilization by allowing other task to run inside allocated resources, CPU1 in this case, via backfill.
+
+![backfill-use-case](../images/backfill-use-case.png)
+
+### Starvation Prevention
```

Comment:
Please separate `reservation` out as a single feature to avoid starvation; we will enable it by default, but backfill will be disabled by default.

Reply:
Starvation prevention and backfill are related. I feel that it is clearer to address both issues together in one design doc.

The implementation, however, can be separated. We will raise two PRs. One for starvation prevention and one for backfill.

Reply:
hm... they're related, but two standalone features: 1. reservation is used to resolve starvation; 2. because of reservation, we need backfill to improve the resource utilization. I'd like to get reservation (starvation prevention) done firstly, and then backfill.